# Homework 1 Linear Equation System Solution Computation

Student Name: 章帅

Student ID: 12163186

## Requirement

- Compute solution $x$ to a square linear equation system $Kx=b$, where $K$ has a dimension $k$.
- Test on at least three difference dimensions $k$ of $K$, at least one value of $k$ is larger than 10000;
- Using at least two difference numerical approaches, with performance analyzed and compared;
- Application on pratical problems is a plus;
- Code submitted together

## Report and code

In [1]:
import numpy as np
import time

### method

#### Steepest descent method

In [2]:
def SD(A, b, x0, max_iter=100000, eps=0.01, res_updt_step=False):
    s = time.time()
    k = 0
    r = b - A @ x0
    delta = r @ r
    x = x0
    while k < max_iter and delta > eps ** 2:
        q = A @ r
        alpha = delta / (r @ q)
        x = x + alpha * r
        if res_updt_step:
            r = r - alpha * q
        else:
            r = b - A @ x
        delta = r @ r
        k = k + 1
    du = time.time() - s
    
    print("=========Steepest descent method result:=========")
    print("solution x:")
    print(x)
    print()
    
    print(f"error: {np.sqrt(delta)}s")
    print()
    
    print(f"duration: {du}s")
    print()

    print(f"number of iteration: {k}")
    
    return x, du, k

#### Conjugate gradient method

In [3]:
def CG(A, b, x0, max_iter=100000, eps=0.01, res_updt_step=False):
    s = time.time()
    k = 0
    r = b - A @ x0
    d = r
    delta = r @ r
    x = x0
    while k < max_iter and delta > eps ** 2:
        q = A @ d
        alpha = delta / (d @ q)
        x = x + alpha * d
        if res_updt_step:
            r = r - alpha * q
        else:
            r = b - A @ x
        _delta = r @ r
        beta = _delta / delta
        delta = _delta
        d = r + beta * d
        k = k + 1
    du = time.time() - s
    
    print("=========Conjugate gradient method result:=========")
    print("solution x:")
    print(x)
    print()

    print(f"error: {np.sqrt(delta)}s")
    print()
    
    print(f"duration: {du}s")
    print()
    
    print(f"number of iteration: {k}")
    
    return x, du, k

### Data simulation and solutions

In [4]:
def generate(dim):
    Q = np.random.randn(dim, dim)
    A = np.transpose(Q) @ Q
    x_opt = np.random.randn(dim)
    b = A @ x_opt
    print(f"=========== Equation Ax = b of dimension {dim} where ===========")
    print("A:")
    print(A)
    print()
    print("b:")
    print(b)
    print()
    print("optimal x:")
    print(x_opt)
    return A, b, x_opt

#### dimension = 4

In [5]:
dim = 4
x0 = np.random.randn(dim)
A, b, x_opt = generate(dim)

=========== Equation Ax = b of dimension 4 where ===========
A:
[[ 4.33761631  0.67135742  3.19667549  0.55639509]
 [ 0.67135742  1.17181057 -0.77912114 -1.37109757]
 [ 3.19667549 -0.77912114  7.69403839  3.79526131]
 [ 0.55639509 -1.37109757  3.79526131  4.1727923 ]]

b:
[ 2.46219584  1.37094707 -0.68671998 -2.93401428]

optimal x:
[ 0.68788238 -0.29138878  0.06302457 -0.94791821]


In [6]:
x, du, num_iter = SD(A, b, x0)

=========Steepest descent method result:=========
solution x:
[ 0.68429859 -0.27624406  0.06259922 -0.94191288]

error: 0.009016234514821505s

duration: 0.00018978118896484375s

number of iteration: 27


In [7]:
x, du, num_iter = CG(A, b, x0)

=========Conjugate gradient method result:=========
solution x:
[ 0.68788238 -0.29138878  0.06302457 -0.94791821]

error: 2.334829376188003e-13s

duration: 0.00015616416931152344s

number of iteration: 4


#### dimension = 100

In [8]:
dim = 100
x0 = np.random.randn(dim)
A, b, x_opt = generate(dim)

=========== Equation Ax = b of dimension 100 where ===========
A:
[[ 85.22788763 -12.03193864  12.96167702 ...  20.74913119  -3.83422277
  -10.33375485]
 [-12.03193864  97.48711323   7.47371039 ...   8.11689252  21.7930118
   -2.0852225 ]
 [ 12.96167702   7.47371039  86.74749111 ...  10.44109497  -1.20299817
   -1.68676335]
 ...
 [ 20.74913119   8.11689252  10.44109497 ...  92.44284532  19.18011377
  -11.17580185]
 [ -3.83422277  21.7930118   -1.20299817 ...  19.18011377  77.22635908
    0.56729855]
 [-10.33375485  -2.0852225   -1.68676335 ... -11.17580185   0.56729855
  113.87270714]]

b:
[ 1.29245459e+02  5.66668981e+01  1.00720168e+02  5.50732505e+01
 -7.32765547e+01 -4.78521111e+01  4.96202433e+01 -8.97033721e+01
 -3.83472217e+01 -1.23647993e+02  4.95748403e+02 -1.26263468e+02
  2.05583130e+02 -1.54081159e+02  1.41123538e+02  6.11611649e+01
 -4.46682854e+01 -2.76549919e+02 -1.08194055e+02 -1.39395833e+02
  9.21231873e+01  7.35394598e-02  1.42741633e+02 -1.17072683e+02
 -7.20703423e

In [9]:
x, du, num_iter = SD(A, b, x0)

=========Steepest descent method result:=========
solution x:
[ 2.66616029  0.86086208  0.51591788  1.00440374 -0.5279931  -1.02561181
  0.69492732 -0.06172791 -1.12832021 -0.42602526  1.30584894 -0.68254223
  1.43302966 -1.02136121  1.53536788  0.20547215 -0.62475982 -0.76560485
 -1.00720372 -0.49948502 -0.14640663 -0.59995362  0.80922831 -0.56878546
 -0.54879943  0.69598636 -1.09387701 -0.9461951  -0.88233959 -0.16644711
 -1.16649521 -1.99227875  1.70061432 -0.18610805 -0.42271276  1.19691124
  0.28786798  0.34586644 -1.44037575  1.527155    0.46028098 -0.06866398
 -1.16340915 -1.47504758  0.81066506 -1.62036093 -0.93961307 -0.09248625
  0.2139858   0.4950873   0.27393065  1.47818753 -0.03122918 -0.65080288
 -1.09202114 -0.20860463 -0.58583903 -0.70440593  0.32427492  0.72860034
 -0.30003468 -0.14257105  0.22911095  0.4897119   1.25399769  1.32811828
  0.19901199  0.68791696  0.95931061 -1.01605582 -0.26461962  1.1776962
 -0.25716704 -0.78407353  2.01319791 -0.19555604 -0.07029033 -0

In [10]:
x, du, num_iter = CG(A, b, x0)

=========Conjugate gradient method result:=========
solution x:
[ 2.69421728  0.86683495  0.48933663  0.97858067 -0.50507983 -1.04032517
  0.69784627 -0.03977556 -1.13286025 -0.42098714  1.29763832 -0.69644586
  1.44966441 -0.96685131  1.55112143  0.24704893 -0.61591289 -0.78773557
 -1.02787325 -0.48138194 -0.12197101 -0.61423905  0.79892734 -0.58045928
 -0.53883257  0.68655672 -1.10401067 -0.95754603 -0.89783605 -0.19216239
 -1.17893673 -1.98476201  1.68097612 -0.20293491 -0.42932755  1.19850716
  0.30035141  0.36938418 -1.44480637  1.52037447  0.47320656 -0.0947396
 -1.16720854 -1.4702824   0.8129526  -1.61129043 -0.91779646 -0.11691986
  0.21485126  0.5221659   0.29401424  1.47467088 -0.04030315 -0.64858859
 -1.09335028 -0.22772519 -0.57896508 -0.70195343  0.35239841  0.71449728
 -0.27996535 -0.12831163  0.2532717   0.47343261  1.26823485  1.31722485
  0.21369705  0.69736988  0.97440045 -1.02730504 -0.2835342   1.16550542
 -0.25040795 -0.7838458   2.01400321 -0.20399203 -0.07608085 

#### dimension =12000

In [11]:
dim = 12000
x0 = np.random.randn(dim)
A, b, x_opt = generate(dim)

=========== Equation Ax = b of dimension 12000 where ===========
A:
[[ 1.21880633e+04  9.88698980e+01 -2.32272079e+02 ... -1.16943679e+00
  -5.91459064e+01  6.55435382e+01]
 [ 9.88698980e+01  1.21378518e+04  3.28215030e+01 ... -2.09049302e+01
  -2.19573031e+02  5.97700265e+01]
 [-2.32272079e+02  3.28215030e+01  1.17361630e+04 ...  1.32651383e+01
  -4.71772138e+00 -1.70977419e+02]
 ...
 [-1.16943679e+00 -2.09049302e+01  1.32651383e+01 ...  1.17477501e+04
   4.63156191e+01  8.14606780e+01]
 [-5.91459064e+01 -2.19573031e+02 -4.71772138e+00 ...  4.63156191e+01
   1.18719427e+04 -2.77923791e+02]
 [ 6.55435382e+01  5.97700265e+01 -1.70977419e+02 ...  8.14606780e+01
  -2.77923791e+02  1.18936448e+04]]

b:
[ 20974.10857459 -11654.11625521 -13556.30503251 ...  -7901.8189511
 -37319.90549283 -23635.2284053 ]

optimal x:
[ 1.41004036 -0.88183657 -0.25392586 ...  0.87416762 -1.04718924
 -0.49992993]


In [13]:
x, sd_du, sd_num_iter = SD(A, b, x0, res_updt_step=True)

=========Steepest descent method result:=========
solution x:
[ 1.50794937 -0.88099635 -0.28149596 ...  0.93957333 -1.09205439
 -0.56997257]

error: 0.9628189127488034s

duration: 3489.171194791794s

number of iteration: 100000


In [12]:
x, cg_du, cg_num_iter = CG(A, b, x0, res_updt_step=True)

=========Conjugate gradient method result:=========
solution x:
[ 1.4087051  -0.90551614 -0.26047164 ...  0.86160935 -1.04348693
 -0.4974341 ]

error: 0.009984484182377637s

duration: 507.076691865921s

number of iteration: 13634


### Performance comparison

- Theoretically, the conjugate gradient approach has a superlinear convergence speed, i.e. $\lim_{n \to \infty} \frac{||x^{n+1} - x^{\star}||}{||x^{n} - x^{\star}||}=0$, outperforming the steepest gradient descent approach, which finds the optimal solution linearly, i.e. $\lim_{n \to \infty} \frac{||x^{n+1} - x^{\star}||}{||x^{n} - x^{\star}||}=\lambda > 0$
- Experimentally, from the results above, we can see the conjugate gradient method uses less steps and finds the optimal solution much quicker within a given error tolerance i.e. $||Ax - b|| < 0.01$ than the steepest gradient descent method.
| Method | Equation dimension | Time cost (s) | Number of iteration | Error |
| --- | --- | --- |  |  |
| Steepest descent   | 4     | 0.00019 | 27                     | 0.0091  |
| Conjugate gradient | 4     | 0.00016 | 4                      | 2.3e-13 |
| Steepest descent   | 100   | 0.066   | 4747                   |  0.010  |
| Conjugate gradient | 100   | 0.018   | 142                    |  0.0031 |
| Steepest descent   | 12000 | 3489    | 100000 (max iteration) |  0.96   |
| Conjugate gradient | 12000 | 507     | 13634                  |  0.010  |
